File: app/android/src/main/kotlin/io/gratia/app/sensors/WifiManager.kt
package io.gratia.app.sensors

import android.Manifest
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.net.wifi.WifiManager as AndroidWifiManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.core.content.ContextCompat
import java.security.MessageDigest

/**
 * Wi-Fi manager for Proof of Life network detection.
 *
 * Detects visible Wi-Fi networks by triggering periodic scans and reading
 * scan results. Each BSSID (access point MAC address) is hashed to an opaque
 * long value before being sent across the FFI boundary. The Rust PoL engine
 * uses the count of distinct BSSID hashes to satisfy the "connected to at
 * least one Wi-Fi network" parameter.
 *
 * PRIVACY: Raw BSSIDs are never stored or transmitted. Only SHA-256
 * truncated hashes are forwarded.
 *
 * WHY Wi-Fi scanning interval is 20 minutes: Android throttles Wi-Fi scans
 * to 4 per 2-minute period in the foreground and 1 per 30 minutes in the
 * background (Android 9+). Our 20-minute interval stays well within the
 * background throttle limit.
 */
class GratiaWifiManager(
    private val context: Context,
    private val listener: SensorEventListener
) {

    companion object {
        private const val TAG = "GratiaWifi"

        // WHY: 20-minute interval stays within Android's background scan throttle
        // (1 scan per 30 minutes on Android 9+) with margin. Even a single
        // successful scan satisfies the Wi-Fi PoL parameter.
        private const val SCAN_INTERVAL_MS = 20L * 60 * 1000 // 20 minutes
    }

    private val handler = Handler(Looper.getMainLooper())
    private val androidWifiManager: AndroidWifiManager? =
        context.applicationContext.getSystemService(Context.WIFI_SERVICE) as? AndroidWifiManager

    private val digest: MessageDigest = MessageDigest.getInstance("SHA-256")
    private var isRunning = false
    private var receiverRegistered = false

    // BroadcastReceiver for scan results.
    private val scanResultsReceiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context, intent: Intent) {
            if (intent.action == AndroidWifiManager.SCAN_RESULTS_AVAILABLE_ACTION) {
                val success = intent.getBooleanExtra(
                    AndroidWifiManager.EXTRA_RESULTS_UPDATED, false
                )
                processScanResults(success)
            }
        }
    }

    private val scanRunnable = object : Runnable {
        override fun run() {
            if (!isRunning) return
            triggerScan()
            handler.postDelayed(this, SCAN_INTERVAL_MS)
        }
    }

    /**
     * Start periodic Wi-Fi scanning.
     *
     * Requires ACCESS_FINE_LOCATION (or ACCESS_COARSE_LOCATION on older APIs)
     * to read scan results. Gracefully handles missing permissions.
     */
    fun start() {
        if (isRunning) return

        if (androidWifiManager == null) {
            Log.w(TAG, "WifiManager not available")
            return
        }

        if (!hasWifiPermission()) {
            Log.w(TAG, "Wi-Fi scan permission not granted — Wi-Fi PoL parameter may not be met")
            return
        }

        // Register receiver for scan results.
        val filter = IntentFilter(AndroidWifiManager.SCAN_RESULTS_AVAILABLE_ACTION)
        context.registerReceiver(scanResultsReceiver, filter)
        receiverRegistered = true

        isRunning = true

        // Read current scan results immediately (may have cached results).
        processScanResults(cached = true)

        // Schedule periodic scans.
        handler.postDelayed(scanRunnable, SCAN_INTERVAL_MS)

        Log.i(TAG, "Wi-Fi scanning started (interval=${SCAN_INTERVAL_MS / 60000}min)")
    }

    /** Stop scanning and unregister receivers. */
    fun stop() {
        if (!isRunning) return

        isRunning = false
        handler.removeCallbacks(scanRunnable)

        if (receiverRegistered) {
            try {
                context.unregisterReceiver(scanResultsReceiver)
            } catch (e: IllegalArgumentException) {
                // Receiver was already unregistered.
            }
            receiverRegistered = false
        }

        Log.i(TAG, "Wi-Fi scanning stopped")
    }

    /** Check whether this manager is actively scanning. */
    fun isActive(): Boolean = isRunning

    /** Check whether Wi-Fi hardware is present. */
    fun isAvailable(): Boolean = androidWifiManager != null

    // ========================================================================
    // Internal
    // ========================================================================

    @Suppress("DEPRECATION")
    private fun triggerScan() {
        if (!hasWifiPermission()) return

        // WHY: startScan() is deprecated in Android 9+ and throttled, but it's
        // still the only way to request a fresh scan. We also process cached
        // results from getScanResults() which works reliably.
        try {
            androidWifiManager?.startScan()
            Log.d(TAG, "Wi-Fi scan triggered")
        } catch (e: SecurityException) {
            Log.w(TAG, "SecurityException triggering Wi-Fi scan: ${e.message}")
        }
    }

    private fun processScanResults(success: Boolean = true, cached: Boolean = false) {
        if (!hasWifiPermission()) return

        try {
            @Suppress("MissingPermission")
            val results = androidWifiManager?.scanResults ?: return

            if (results.isEmpty()) {
                Log.d(TAG, "Wi-Fi scan completed — no networks found")
                return
            }

            val bssidHashes = results.mapNotNull { result ->
                val bssid = result.BSSID ?: return@mapNotNull null
                hashBssid(bssid)
            }.distinct()

            if (bssidHashes.isNotEmpty()) {
                val source = if (cached) "cached" else if (success) "fresh" else "stale"
                Log.d(TAG, "Wi-Fi scan ($source) — ${bssidHashes.size} networks found")
                listener.onWifiScan(bssidHashes)
            }
        } catch (e: SecurityException) {
            Log.w(TAG, "SecurityException reading Wi-Fi scan results: ${e.message}")
        }
    }

    /**
     * Hash a BSSID (MAC address) to an opaque 8-byte long.
     *
     * Same hashing approach as BluetoothManager for consistency.
     */
    private fun hashBssid(bssid: String): Long {
        val hashBytes = digest.digest(bssid.toByteArray(Charsets.UTF_8))
        digest.reset()

        var result = 0L
        for (i in 0 until 8) {
            result = result or ((hashBytes[i].toLong() and 0xFF) shl (i * 8))
        }
        return result
    }

    private fun hasWifiPermission(): Boolean {
        // WHY: ACCESS_FINE_LOCATION is required to get Wi-Fi scan results on
        // Android 8.0+. ACCESS_COARSE_LOCATION is insufficient for BSSID data.
        return ContextCompat.checkSelfPermission(
            context, Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
    }
}

File: app/ios/Gratia/Sources/Sensors/WifiManager.swift
import Foundation
import NetworkExtension
import SystemConfiguration.CaptiveNetwork
import CryptoKit

// ============================================================================
// WifiManager — Wi-Fi network detection for PoL attestation
//
// Mirrors Android's WifiManager.kt:
// - Collects Wi-Fi network BSSID hashes (privacy-preserving)
// - Posts WifiScan sensor events
//
// WHY NEHotspotHelper vs CNCopyCurrentNetworkInfo:
// - CNCopyCurrentNetworkInfo is deprecated in iOS 14+ and requires
//   entitlements that are difficult to obtain.
// - NEHotspotHelper requires the com.apple.developer.networking.HotspotHelper
//   entitlement (granted by Apple on request).
// - As a fallback, we use NEHotspotNetwork.fetchCurrent() (iOS 14+) which
//   works with the standard Access WiFi Information entitlement.
//
// PRIVACY: Raw BSSIDs are never stored or transmitted. They are immediately
// hashed on-device and only the hash is forwarded.
// ============================================================================

final class WifiManager: ObservableObject {

    private let logger = GratiaLogger(tag: "GratiaWifiManager")

    private var isRunning = false
    private var scanTimer: Timer?

    /// Callback for PoL sensor events.
    var onWifiScan: (([UInt64]) -> Void)?

    // WHY: 15-minute interval matches the GPS scan interval. Wi-Fi network
    // identity changes less frequently than Bluetooth environments, so fewer
    // scans are needed. One detected network per day satisfies the PoL
    // "connected to at least one Wi-Fi network" parameter.
    private static let scanIntervalSeconds: TimeInterval = 15 * 60 // 15 minutes

    /// Start periodic Wi-Fi network scanning.
    func start() {
        guard !isRunning else { return }

        isRunning = true
        performScan()
        schedulePeriodic()
        logger.info("Wi-Fi scanning started (interval=\(Int(Self.scanIntervalSeconds / 60))min)")
    }

    /// Stop scanning.
    func stop() {
        guard isRunning else { return }

        scanTimer?.invalidate()
        scanTimer = nil
        isRunning = false
        logger.info("Wi-Fi scanning stopped")
    }

    /// Whether this manager is actively scanning.
    var isActive: Bool { isRunning }

    // MARK: - Private

    private func schedulePeriodic() {
        scanTimer?.invalidate()
        scanTimer = Timer.scheduledTimer(withTimeInterval: Self.scanIntervalSeconds, repeats: true) { [weak self] _ in
            self?.performScan()
        }
    }

    private func performScan() {
        // WHY: iOS 14+ API for fetching the current Wi-Fi network.
        // This is the most reliable method that works with standard entitlements.
        if #available(iOS 14.0, *) {
            NEHotspotNetwork.fetchCurrent { [weak self] network in
                guard let self = self else { return }

                if let network = network {
                    let bssidHash = self.hashBssid(network.bssid)
                    self.logger.debug("Wi-Fi scan: SSID=\(network.ssid), BSSID hash captured")
                    self.onWifiScan?([bssidHash])
                } else {
                    // WHY: nil means either Wi-Fi is off, the device is not connected
                    // to any network, or the entitlement is missing. This is non-fatal;
                    // the PoL network parameter can also be satisfied by Bluetooth.
                    self.logger.debug("Wi-Fi scan: no current network")
                }
            }
        } else {
            // Fallback for older iOS versions using CNCopyCurrentNetworkInfo
            fetchViaLegacyApi()
        }
    }

    /// Legacy fallback using CNCopyCurrentNetworkInfo (iOS < 14).
    private func fetchViaLegacyApi() {
        guard let interfaces = CNCopySupportedInterfaces() as? [String] else {
            logger.debug("Wi-Fi scan: no supported interfaces")
            return
        }

        var hashes: [UInt64] = []

        for interface in interfaces {
            guard let info = CNCopyCurrentNetworkInfo(interface as CFString) as? [String: Any],
                  let bssid = info[kCNNetworkInfoKeyBSSID as String] as? String else {
                continue
            }
            hashes.append(hashBssid(bssid))
        }

        if hashes.isEmpty {
            logger.debug("Wi-Fi scan: no networks detected (legacy API)")
        } else {
            logger.debug("Wi-Fi scan: \(hashes.count) network(s) detected")
            onWifiScan?(hashes)
        }
    }

    /// Hash a BSSID string to an opaque 8-byte UInt64.
    ///
    /// PRIVACY: Raw BSSIDs are never stored or sent. The hash is a one-way
    /// transformation for detecting "same network seen again".
    private func hashBssid(_ bssid: String) -> UInt64 {
        let data = Data(bssid.utf8)
        let hash = SHA256.hash(data: data)

        var result: UInt64 = 0
        let bytes = Array(hash)
        for i in 0..<8 {
            result |= UInt64(bytes[i]) << (i * 8)
        }
        return result
    }
}
